src/precomputed_quantities.jl

```diff
@@ -319,18 +370,238 @@ function set_precomputed_quantities!(Y, p, t)
+                h_tot_level = Fields.field_values(Fields.level(ᶜh_tot, i))
+                q_tot_level = Fields.field_values(Fields.level(q_tot, i))
+                p_level = Fields.field_values(Fields.level(ᶜp, i))
+                Φ_level = Fields.field_values(Fields.level(ᶜΦ, i))
+
+                local_geometry_level = Fields.field_values(
+                    Fields.level(Fields.local_geometry_field(Y.c), i),
+                )
+                local_geometry_halflevel = Fields.field_values(
+                    Fields.level(Fields.local_geometry_field(Y.f), i - half),
+                )
+                ∂x∂ξ_level = local_geometry_level.∂x∂ξ.components.data
+                end_index = fieldcount(eltype(∂x∂ξ_level)) # This will be 4 in 2D and 9 in 3D.
+                ∂x³∂ξ³_level = ∂x∂ξ_level.:($end_index)
+
+                ρaʲ_prev_level = Fields.field_values(Fields.level(ᶜρaʲ, i - 1))
+                u³ʲ_prev_halflevel =
+                    Fields.field_values(Fields.level(ᶠu³ʲ, i - 1 - half))
+                u³ʲ_data_prev_halflevel = u³ʲ_prev_halflevel.components.data.:1
+                h_totʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜh_totʲ, i - 1))
+                q_totʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜq_totʲ, i - 1))
+                ρʲ_prev_level = Fields.field_values(Fields.level(ᶜρʲ, i - 1))
+                ρ_ref_prev_level =
+                    Fields.field_values(Fields.level(ᶜρ_ref, i - 1))
+                ∇p_perturb³_prev_level =
+                    Fields.field_values(Fields.level(ᶜ∇p_perturb³, i - 1))
+                ∇p_perturb³_prev_level_data =
+                    ∇p_perturb³_prev_level.components.data.:1
+                ∇Φ³_prev_level = Fields.field_values(Fields.level(ᶜ∇Φ³, i - 1))
+                ∇Φ³_prev_level_data = ∇Φ³_prev_level.components.data.:1
+
+                local_geometry_prev_level = Fields.field_values(
+                    Fields.level(Fields.local_geometry_field(Y.c), i - 1),
+                )
+                local_geometry_prev_halflevel = Fields.field_values(
+                    Fields.level(
+                        Fields.local_geometry_field(Y.f),
+                        i - 1 - half,
+                    ),
+                )
+
+                @. ρaʲu³ʲ_data =
+                    (1 / local_geometry_halflevel.J) * (
+                        local_geometry_prev_halflevel.J *
+                        ρaʲ_prev_level *
+                        u³ʲ_data_prev_halflevel
+                    )
+
+                @. ρaʲu³ʲ_datau³ʲ_data =
+                    (1 / local_geometry_halflevel.J^2) * (
+                        local_geometry_prev_halflevel.J^2 *
+                        ρaʲ_prev_level *
+                        u³ʲ_data_prev_halflevel *
+                        u³ʲ_data_prev_halflevel
+                    )
+
+                @. ρaʲu³ʲ_datau³ʲ_data -=
+                    (1 / local_geometry_halflevel.J^2) * (
```

Comment:
I was trying to follow along with the overleaf equations. It would be great to add some documentation to ClimaAtmos that explains the operators. For example, I got confused about where the `g^3l` term is?

Reply:
Yes! I will add some documentation next week. The `g^3l` term is in the project e.g. [here](https://github.com/CliMA/ClimaAtmos.jl/blob/a6ed6d10ea53beb91ae56482ca3790d0a60b12e0/src/cache/precomputed_quantities.jl#L381).